Runtime/Scripts/PersistentConnection.cs

```diff
@@ -131,192 +217,193 @@ IEnumerator OnIntervalEdgeEventsLocation(EdgeEventsConnection connection)
+          yield break;
+        }
+      }
+      yield return new WaitForSecondsRealtime(config.latencyConfig.updateIntervalSeconds);
+      yield return StartCoroutine(LocationService.EnsureLocation());
+      if (hasTCPPorts)
+      {
+        connection.TestConnectAndPostLatencyUpdate(host, (uint)config.latencyTestPort, location).ConfigureAwait(false);
+      }
+      else
+      {
+        integration.matchingEngine.EdgeEventsConnection
+          .TestPingAndPostLatencyUpdate
+          (integration.GetHost(), location).ConfigureAwait(false);
+      }
+      latencyUpdatesCounter++;
+      yield return StartCoroutine(OnIntervalEdgeEventsLatency(connection, host));
+    }
+
+    async void HandleReceivedEvents(ServerEdgeEvent edgeEvent)
+    {
+      Logger.Log("Received event type: " + edgeEvent.EventType);
+      FindCloudletEventTrigger trigger = event_trigger_dict[edgeEvent.EventType];
+      if (config.newFindCloudletEventTriggers.Contains(trigger))
+      {
+        switch (trigger)
+        {
+          case FindCloudletEventTrigger.Error:
+            PropagateError(FindCloudletEventTrigger.Error, edgeEvent.ErrorMsg);
+            return;
+          case FindCloudletEventTrigger.LatencyTooHigh:
+            ProcessLatency(edgeEvent.Statistics);
+            break;
+          default:
+            if (edgeEvent.NewCloudlet != null)
             {
-              latencyUpdatesRunning = false;
-              yield break;
+              PropagateSuccess(trigger, edgeEvent.NewCloudlet);
             }
-          }
-          yield return new WaitForSecondsRealtime(config.latencyConfig.updateIntervalSeconds);
-          yield return StartCoroutine(LocationService.EnsureLocation());
+            return;
+        }
+      }
+      else
+      {
+        if (edgeEvent.EventType == ServerEventType.EventLatencyRequest)
+        {
+          string host = integration.GetHost();
+          Loc location = LocationService.RetrieveLocation();
           if (hasTCPPorts)
           {
-            connection.TestConnectAndPostLatencyUpdate(host, (uint)config.latencyTestPort, location).ConfigureAwait(false);
+            await integration.matchingEngine.EdgeEventsConnection.TestConnectAndPostLatencyUpdate(host, (uint)config.latencyTestPort, location);
           }
           else
           {
-            integration.matchingEngine.EdgeEventsConnection
-              .TestPingAndPostLatencyUpdate
-              (integration.GetHost(), location).ConfigureAwait(false);
+            await integration.matchingEngine.EdgeEventsConnection.TestPingAndPostLatencyUpdate(host, location);
           }
-          latencyUpdatesCounter++;
-          yield return StartCoroutine(OnIntervalEdgeEventsLatency(connection, host));
         }
+      }
+    }
 
-    async void HandleServerReceivedEvents(ServerEdgeEvent edgeEvent)
+    async void ProcessLatency(Statistics stats)
+    {
+      FindCloudletReply currentCloudlet = integration.latestFindCloudletReply;
```

Comment:
Ah, so using the latestFindCloudletReply was an issue that Garner and I resolved in https://mobiledgex.atlassian.net/browse/EDGECLOUD-5100. Since, you are setting latestFindCloudletReply after every FindCloudlet call, there is a case where the FindCloudletPerformance call returns a cloudlet that we will not use because it is not better than the performanceMargin. Then the latestFindCloudletReply will be set to a cloudlet that the client is not actually on. So, if the client actually moves closer to that cloudlet in the future, it will compare it to latestFindCloudletReply (which will be the same) and not switch to it when it is supposed to.

The fix we made was to keep track of currentFindCloudletReply as an EdgeEventsConnection class variable (or PersistentConnection in your case) and that will be the current cloudlet that we are receiving events about.

Reply:
has this been fixed?

Reply:
Yes I am passing the currentCloudlet to CompareLatencies to revert to old cloudlet if the new cloudlet is not better.
Let me know if this is suffcient.